linux/src/threadpool_linux.c

```diff
@@ -217,29 +194,25 @@ static int threadpool_work_func(void* param)
                     if (work_function != NULL)
                     {
                         work_function(work_function_ctx);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_039: [ threadpool_work_func shall acquire the shared SRW lock by calling srw_lock_acquire_shared. ]*/
+
                         srw_lock_acquire_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_040: [ threadpool_work_func shall save pending_work_item_count_ptr is not NULL in is_pending_work_item_count_ptr_not_null. ]*/
                         bool is_pending_work_item_count_ptr_not_null = (NULL != threadpool->task_array[current_index].pending_work_item_count_ptr);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_041: [ threadpool_work_func shall release the shared SRW lock by calling srw_lock_release_shared. ]*/
                         srw_lock_release_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_042: [ If the is_pending_work_item_count_ptr_not_null is TRUE then: ]*/
+
                         if (is_pending_work_item_count_ptr_not_null)
-                        {
-                            /* Codes_SRS_THREADPOOL_LINUX_05_043: [ threadpool_work_func shall acquire the exclusive SRW lock by calling srw_lock_acquire_exclusive. ]*/
+                        {                            
                             srw_lock_acquire_exclusive(threadpool->srw_lock);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_044: [ threadpool_work_func shall decrement the pending_work_item_count_ptr by calling interlocked_decrement. ]*/
-                            interlocked_decrement(threadpool->task_array[current_index].pending_work_item_count_ptr);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_045: [ threadpool_work_func shall send wake up signal to single listener for the address in pending_work_item_count_ptr. ]*/
-                            wake_by_address_single(threadpool->task_array[current_index].pending_work_item_count_ptr);
-                            /* Codes_SRS_THREADPOOL_LINUX_05_046: [ threadpool_work_func shall release the shared SRW lock by calling srw_lock_release_exclusive. ]*/
+                            if (interlocked_decrement(threadpool->task_array[current_index].pending_work_item_count_ptr) == 0)
+                            {
+                                wake_by_address_single(threadpool->task_array[current_index].pending_work_item_count_ptr);                         
+                            }
                             srw_lock_release_exclusive(threadpool->srw_lock);
                         }
```

Comment:
I think this acquire shared/release shared and then acquire exclusive and release exclusive is useless.

In line 199 we evaluate if threadpool->task_array[current_index].pending_work_item_count_ptr is NULL, but after the lock is released that can easily change.

I think this whole dance of pending_work_item_count_ptr is just a poor substitute for having a field here with a THANDLE of the WORK_ITEM. After all that is what we aim at - making sure that a reference is kept to the THANDLE(WORK_ITEM) while it still has to execute.

Reply:
Hi Dan, it won't change in this case since it is assigned in threadpool_schedule_work_item for an index in thread_array and hence not NULL for a given 'index' and won't change after the shared lock is released. When threadpool_schedule_work is used, it will be NULL for the given 'index' and won't change after shared lock is released.

This is just a code switch until threadpool_schedule_,work goes away